pennylane/labs/dla/dense_util.py

```diff
@@ -0,0 +1,456 @@
+    >>> trace_inner_product(qml.X(0) + qml.Y(0), qml.Y(0) + qml.Z(0))
+    1.0
+
+    """
+    if getattr(A, "pauli_rep", None) is not None and getattr(B, "pauli_rep", None) is not None:
+        return (A.pauli_rep @ B.pauli_rep).trace()
+
+    if not isinstance(A, type(B)):
+        raise TypeError("Both input operators need to be of the same type")
+
+    if isinstance(A, np.ndarray):
+        assert A.shape[-2:] == B.shape[-2:]
+        # The axes of the first input are switched, compared to tr[A@B], because we need to
+        # transpose A.
+        return np.tensordot(A.conj(), B, axes=[[-1, -2], [-1, -2]]) / A.shape[-1]
+
+    if isinstance(A, (PauliSentence, PauliWord)):
+        return (A @ B).trace()
+
+    raise NotImplementedError
+
+
+def adjvec_to_op(adj_vecs, basis, is_orthogonal=True):
+    r"""Transform adjoint vector representations back into operator format.
+
+    This function simply reconstructs :math:`\hat{O} = \sum_j c_j \hat{b}_j` given the adjoint vector
+    representation :math:`c_j` and basis :math:`\hat{b}_j`.
+
+    .. seealso:: :func:`~op_to_adjvec`
+
+    Args:
+        adj_vecs (np.ndarray): collection of vectors with shape ``(batch, len(basis))``
+        basis (List[Union[PauliSentence, Operator, np.ndarray]]): collection of basis operators
+        is_orthogonal (bool): Whether the ``basis`` consists of orthogonal elements.
+
+    Returns:
+        list: collection of operators corresponding to the input vectors read in the input basis.
+        The operators are in the format specified by the elements in ``basis``.
+
+    **Example**
+
+    >>> from pennylane.labs.dla import adjvec_to_op
+    >>> c = np.array([[0.5, 0.3, 0.7]])
+    >>> basis = [qml.X(0), qml.Y(0), qml.Z(0)]
+    >>> adjvec_to_op(c, basis)
+    [0.5 * X(0) + 0.3 * Y(0) + 0.7 * Z(0)]
+
+    """
+    assert qml.math.shape(adj_vecs)[1] == len(basis)
+
+    if all(isinstance(op, PauliSentence) for op in basis):
+        if not is_orthogonal:
+            gram = _gram_ps(basis)
+            adj_vecs = np.tensordot(adj_vecs, np.linalg.pinv(sqrtm(gram)), axes=[[1], [0]])
+        res = []
+        for vec in adj_vecs:
+            op_j = sum(c * op for c, op in zip(vec, basis))
+            op_j.simplify()
+            res.append(op_j)
+        return res
+
+    if all(isinstance(op, Operator) for op in basis):
+        if not is_orthogonal:
+            basis_ps = [op.pauli_rep for op in basis]
+            gram = _gram_ps(basis_ps)
+            adj_vecs = np.tensordot(adj_vecs, np.linalg.pinv(sqrtm(gram)), axes=[[1], [0]])
+        res = []
+        for vec in adj_vecs:
+            op_j = sum(c * op for c, op in zip(vec, basis))
+            op_j = qml.simplify(op_j)
+            res.append(op_j)
+        return res
+
+    if isinstance(basis, np.ndarray) or all(isinstance(op, np.ndarray) for op in basis):
+        if not is_orthogonal:
+            gram = np.tensordot(basis, basis, axes=[[1, 2], [2, 1]]).real / basis[0].shape[0]
```

Comment:
Could do 
```suggestion
            gram = trace_inner_product(basis, basis).real
```
here, I think?